mitmproxy/net/http/request.py

```diff
@@ -367,13 +367,64 @@ def _get_urlencoded_form(self):
                 pass
         return ()
 
+    def _sanitized_form_data_replacement(self, form_data):
+        # check if unmodified content changes with URL encoding
+        url_decoded_content = urllib.parse.parse_qsl(self.content, keep_blank_values=True)
+        url_reencoded_content = urllib.parse.urlencode(url_decoded_content, doseq=True).encode()
+        if self.content == url_reencoded_content:
+            # encoding the unmodified content doesn't change anything, so we should be good to go on as usual
+            self.content = mitmproxy.net.http.url.encode(form_data).encode()
+        else:
+            # if we are here, we have to alter encoding behavior to mimic the one used in unmodified POST request
+
+            # check if content isn't encoded at all
+            if self.content == urllib.parse.unquote(url_reencoded_content.decode()).encode():
+                # if we land here, the new content mustn't be URL encoded thus we decode the result before setting new content
+                # this is the case for "https://www.facebook.com/ajax/updatestatus.php" but additionally the "=" sign is
+                # omitted on SOME empty keys
+                self.content = urllib.parse.unquote(mitmproxy.net.http.url.encode(form_data)).encode()
+            # check if content isn't encoded at all and equal signs are omitted
+            elif self.content == urllib.parse.unquote(url_reencoded_content.decode()).replace("=&", "&").encode():
+                # remove equal sign for keys without value
+                self.content = urllib.parse.unquote(mitmproxy.net.http.url.encode(form_data)).replace("=&", "&").encode()
```

Comment:
Line 387 is the check for "special case 2" is similar to "special case 1". Beside urldecoding (same principle) our test data again, all occurrences of "=&" are replaced with "&".

So this special case check could be rewritten to:
`if self.content == urldecode(urlencode(urldecode(self.content))).replace("=&", "&"):`
Which resolves to:
`if self.content == urldecode(self.content).replace("=&", "&"):`

In words: if umodified content isn't encoded and empty keys don't have an equal sign appended.

Line 389 fixes the result of `mitmproxy.net.http.url.encode(form_data)`

This code ignores another corner case:
For `self.content = b'key1&key2` without further data modification `mitmproxy.net.http.url.encode(form_data).encode()` would return  `b'key1=&key2=' `
As the check above only touches "=&" the equal sign tailing key2 wouldn't be replaced. Thus the condition check wouldn't hit, because the `elif` statments ends up comparing:
`b'key1=&key2=' == b'key1&key2='`

So this is a more complex case, which I haven't checked for, but could easily be added in with a new `elif` following the principles of the other special cases.